compiler/rustc_mir_transform/src/lib.rs

```diff
@@ -299,6 +299,7 @@ fn mir_built(tcx: TyCtxt<'_>, def: LocalDefId) -> &Steal<Body<'_>> {
             // by-move and by-mut bodies if needed. We do this first so
             // they can be optimized in lockstep with their parent bodies.
             &coroutine::ByMoveBody,
+            &coroutine::relocate_upvars::RelocateUpvars,
```

Comment:
It would be useful to leave a note why this needs to happen after `ByMoveBody`. If we flip these two coroutine passses, we get test failures, right?

Reply:
True and indeed the transformed MIR made no sense because the types of the MIR locals would be wrong.

Reply:
I would also really like to see tests that demonstrate that both closure bodies (regular and by-move) still make sense after this pass

Reply:
> indeed the transformed MIR made no sense because the types of the MIR locals would be wrong.

Yeah, though we could also modify the `ByMoveBody` pass to adjust the types on the MIR locals. But having it in this order keeps it much simpler IMO.